source/ConditionsStore.cpp

```diff
@@ -459,3 +465,60 @@ bool ConditionsStore::VerifyProviderLocation(const string &name, DerivedProvider
 				", because it is within range of prefixed derived provider \"" + ce.provider->name + "\".");
 	return true;
 }
+
+
+
+// Expands the name of a condition to the full name based on the $[] blocks it contains.
+string ConditionsStore::Expand(const string &name) const
+{
+	using size_t = string::size_type;
+	size_t it = name.find("$[");
+
+	// If there is nothing to expand, return.
+	if(it == string::npos)
+		return name;
+
+	const size_t len = name.length();
+	string builder;
+	size_t end = -1; // the end of the previous expansion
+
+	while(it != string::npos)
+	{
+		// add the section between the current expansion and the previous one
+		builder.append(name.substr(end + 1, it - end - 1)); // does not append the iterator char
+		int depth = 1;
+		// find the end of this expansion
+		bool valid = true;
+		for(size_t i = it + 2;; i++)
+		{
+			if(i == len)
+			{
+				// If there is no ] to close this expansion, end the lookup.
+				valid = false;
+				break;
+			}
+			if(i < len - 1 && name[i] == '$' && name[i + 1] == '[')
+			{
+				depth++;
+				i++;
+			}
+			else if(name[i] == ']')
+			{
+				depth--;
+				if(depth == 0)
+				{
+					end = i;
+					break;
+				}
```

Comment:
I am thinking, can't we do this easier by first using a find on the string to see if another [ is coming if not find a ]

Reply:
Yes, but it would be less efficient because it would iterate over the string multiple times.

It could also be used to build stacks of the opening/closing brackets and use that to resolve all expansions, but that seemed too complicated.